veg2hab/functionele_samenhang.py

```diff
@@ -0,0 +1,295 @@
+from collections import defaultdict
+from copy import deepcopy
+from typing import Dict, List, Set, Tuple
+
+import geopandas as gpd
+import pandas as pd
+
+from veg2hab.enums import KeuzeStatus, Kwaliteit
+from veg2hab.habitat import HabitatKeuze
+from veg2hab.io.common import Interface
+
+
+class UnionFind:
+    """
+    Implementatie van de Union-Find/Disjoint-Set datastructuur
+    https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+    https://www.youtube.com/watch?v=ayW5B2W9hfo
+
+    Wordt gebruikt om paren van items te clusteren op basis van gemeenschappelijke elementen
+    [1, 2], [2, 3], [2, 4], [5, 5]          ->  [1, 2, 3, 4], [5]
+    [1, 2], [3, 4], [5, 6], [2, 3]          ->  [1, 2, 3, 4], [5, 6]
+    [1, 2], [2, 3], [4, 5], [1, 6], [4, 7]  ->  [1, 2, 3, 6], [4, 5, 7]
+    """
+
+    def __init__(self):
+        self.parent_mapper = {}
+        self.rank = {}
+
+    def find_root(self, item) -> int:
+        """
+        Doorloopt de boom van item via de parent_mapper en geeft de root terug
+        """
+        if self.parent_mapper[item] != item:
+            self.parent_mapper[item] = self.find_root(self.parent_mapper[item])
+        return self.parent_mapper[item]
+
+    def union(self, item1, item2) -> None:
+        """
+        Combineert de bomen van item1 en item2
+        """
+        root1 = self.find_root(item1)
+        root2 = self.find_root(item2)
+        if root1 != root2:
+            if self.rank[root1] > self.rank[root2]:
+                self.parent_mapper[root2] = root1
+            elif self.rank[root1] < self.rank[root2]:
+                self.parent_mapper[root1] = root2
+            else:
+                self.parent_mapper[root2] = root1
+                self.rank[root1] += 1
+
+    @staticmethod
+    def cluster_pairs(pairs: List[List]) -> List[List]:
```

Comment:
pair moet dus altijd uit 2 elementen bestaan?

Ik zou dan persoonlijk een list[tuple[Any, Any]] netter vinden.

Of ff een assert dat elke pair lengte twee heeft is ook niet verkeerd.

Reply:
Oh ja en het zijn ook tuples (want lists kun je niet hashen en dus niet als dict keys gebruiken)

Reply:
Hij is nu getypehint met de nieuwe FuncSamenhangID NamedTuple. Het zijn er altijd 2 want ze zijn direct afkomstig uit de _left en _right kolommen van een sjoin.